rfcs/0001-fobnail-architecture-change.md

```diff
@@ -0,0 +1,143 @@
+   specified Provisioning Context.
+9. Platform sends RIMs.
+10. Fobnail verifies RIMs and returns status. RIMs are bound to the Provisioning
+    Context.
+11. At this point additional calls may be issued to send more data. This feature
+    is left for future provisioning extensions.
+12. Platform finalizes provisioning by sending special request.
+13. Fobnail verifies all data bound to the Provisioning Context (AIK and RIMs
+    currently) saves it to flash and sends final status.
+14. Platform is provisioned now.
+
+## Remote platform provisioning
+
+During remote platform provisioning Platform Owner is responsible for
+communicating with platform's TPM, verifying TPM (EK certification and AIK
+challenge), receiving and verifying RIMs. Fobnail's sole responsibility is to
+accept configuration provided by Platform Owner which completes provisioning
+process.
+
+## Platform attestation
+
+Since Fobnail Token is a server, attestation must be requested by the platform.
+After plugging Fobnail Token into USB, attestation must start and complete
+within a defined amount of time, otherwise platform will be considered
+untrustworthy.
+
+1. Platform starts attestation by sending signed metadata to Fobnail Token.
+2. Fobnail token verifies metadata and returns Attestation Context ID together
+   with PCR selection and nonce.
+3. Platform does TPM quote with its AIK, Fobnail Token provided PCR selection
+   and nonce. Result sent to Fobnail Token, to appropriate Attestation Context.
+4. Fobnail Token performs Evidence Appraisal and returns trust decision to the
+   Platform.
+5. Fobnail Token unlocks access to keys stored in flash.
+
+# Fobnail Token services
+
+Fobnail services include among others access to protected files and keys. Access
+is granted only to a platform which has passed attestation. Additional, security
+measures could be imposed for finer grained control where each key access would
+have to be granted to each platform.
+
+Secure files can be read, written or deleted. Primary usage for this feature is
+to keep keys which should be possible to read back, e.g. disk encryption keys
+(having Fobnail Token do all crypto on behalf of Platform would result in too
+high performance hit).
+
+Keys, once written can not be retrieved (neither symmetric nor asymmetric). In
+case of symmetric keys Fobnail Token exposes API to encrypt and decrypt data
+using these keys. In case of asymmetric keys Fobnail Token exposes API to
+decrypt or sign data using a private key, and to read public key. Platform is
+responsible for performing cryptographic operations on public key.
+
+Due to Fobnail Token being exposed as a server over (local) network it could be
+accessed by any unprivileged application (just with access to network stack).
+To prevent any application from accessing Fobnail Token services, all requests
+must be signed with AIK which limits access only to application which have
+access to TPM.
```

Comment:
This document seems to lack failure scenarios what happen when assumed above scenarios fail or because of malicious or wrong behavior of user we lost our capabilities. Are we considering any recovery paths or simply assume data is lost?

Reply:
@pietrushnic You mean to recover after AIK loss?

Currently, losing AIK would result in losing access to all data stored in Fobnail as we generate it only once during provisioning and use the same key for attestation. We could add APIs to replace AIK in Fobnail with a new one or use new AIK during each attestation.
Originally, the idea was to generate new AIK for each attestation but TPM sends private key to host (key can be read only when creating it, later it is not possible to read the key), so, we decided to use generate AIK only once.
Some TPMs support `TPM_CreateLoaded` (but most don't) which doesn't expose private key to host. Still we would have to prove somehow that key was created using `TPM_CreateLoaded` and not something else.

Reply:
@pietrushnic With issue opened https://github.com/fobnail/fobnail/issues/44 can we resolve this and [this](https://github.com/fobnail/docs/pull/39#discussion_r921106920) conversation?